pallets/randomness/src/lib.rs

```diff
@@ -0,0 +1,231 @@
+            Ok(Some(InherentError::Other(
+                sp_runtime::RuntimeString::Borrowed("Inherent required to set babe randomness"),
+            )))
+        }
+
+        // The empty-payload inherent extrinsic.
+        fn create_inherent(_data: &InherentData) -> Option<Self::Call> {
+            Some(Call::set_babe_randomness {})
+        }
+
+        fn is_inherent(call: &Self::Call) -> bool {
+            matches!(call, Call::set_babe_randomness { .. })
+        }
+    }
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+        /// This hook checks, on block finalization, that the required inherent was included and clears
+        /// storage to make it necesary to include it in future blocks as well
+        fn on_finalize(_now: BlockNumberFor<T>) {
+            // Ensure the mandatory inherent was included in the block or the block is invalid
+            // We use take() to make sure this is storage is not set for the next block
+            assert!(
+				<InherentIncluded<T>>::take().is_some(),
+				"Mandatory randomness inherent not included; InherentIncluded storage item is empty"
+			);
+        }
+    }
+
+    // Randomness trait
+    impl<T: Config> frame_support::traits::Randomness<T::Hash, BlockNumberFor<T>> for Pallet<T> {
+        /// Uses the BABE randomness of this epoch to generate a random seed that can be used
+        /// for commitments from the last epoch. The provided `subject` MUST have been commited
+        /// AT LEAST during the last epoch for the result of this function to not be predictable
+        ///
+        /// The subject is a byte array that is hashed (to make it a fixed size) and then concatenated with
+        /// the latest BABE randomness. The result is then hashed again to provide the final randomness.
+        fn random(subject: &[u8]) -> (T::Hash, BlockNumberFor<T>) {
+            // If there's randomness available
+            if let Some((babe_randomness, latest_valid_block)) = LatestBabeRandomness::<T>::get() {
+                let hashed_subject = T::Hashing::hash(subject);
+                let mut digest = Vec::new();
+                // Concatenate the latest randomness with the hashed subject
+                digest.extend_from_slice(babe_randomness.as_ref());
+                digest.extend_from_slice(hashed_subject.as_ref());
+                // Hash it
+                let randomness = T::Hashing::hash(digest.as_slice());
+                // Return the randomness for this subject and the latest block for which this randomness is useful
+                // `subject` commitments done after `latest_valid_block` are predictable, and as such MUST be discarded
+                (randomness, latest_valid_block)
+            } else {
+                // If there's no randomness available, return an empty randomness that's invalid for every block
+                // TODO: Check edge case (randomness requested on the first block of the runtime for example)
```

Comment:
If by first block you mean the genesis block, I think it is safe to say that whoever deploys this chain, will be in control of that, and shouldn't shoot themselves in the foot.